### PR TITLE
Better messages for next on a iterator inside for loops

### LINKED ISSUE #102972: Better messages for `next` inside `for` loops
Given the following code: 
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=521fa77822c348df75f29034633aa31d)

```rust
fn main() {
    let mut chars = "HHeelllloo,,  wwoorrlldd!!".chars();
    for c in chars.by_ref() {
        print!("{}", c);
        chars.next(); // Skip next
    }
}
```

The current output is:

```
error[[E0499]](https://doc.rust-lang.org/stable/error-index.html#E0499): cannot borrow `chars` as mutable more than once at a time
 --> src/main.rs:5:9
  |
3 |     for c in chars.by_ref() {
  |              --------------
  |              |
  |              first mutable borrow occurs here
  |              first borrow later used here
4 |         print!("{}", c);
5 |         chars.next(); // Skip next
  |         ^^^^^^^^^^^^ second mutable borrow occurs here
```

Ideally the output should look like:

```
error[[E0499]](https://doc.rust-lang.org/stable/error-index.html#E0499): cannot borrow `chars` as mutable more than once at a time
 --> src/main.rs:5:9
  |
3 |     for c in chars.by_ref() {
  |              --------------
  |              |
  |              first mutable borrow occurs here
  |              first borrow later used here
4 |         print!("{}", c);
5 |         chars.next(); // Skip next
  |         ^^^^^^^^^^^^ second mutable borrow occurs here
 note: a for loop advances the iterator for you. The result is stored in `c`.
 help: if you want to call `next` within the loop, consider `while let`.
```
or similar. The note and/or help are both subject to changes, this is just a basic idea.
We could also conceivably rewrite the error message to expand on what the note says, like so:
```
error[[E0499]](https://doc.rust-lang.org/stable/error-index.html#E0499): cannot call `next` inside a for loop, as this results in two mutable borrows of `chars`.
 --> src/main.rs:5:9
  |
3 |     for c in chars.by_ref() {
  |              --------------
  |              |
  |              first mutable borrow occurs here
  |              first borrow later used here
4 |         print!("{}", c);
5 |         chars.next(); // Skip next
  |         ^^^^^^^^^^^^ second mutable borrow occurs here
 note: a for loop advances the iterator for you. The result is stored in `c`.
 help: if you want to call `next` within the loop, consider `while let`.
```

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #102972

### CODE REVIEW

**Review by compiler-errors — COMMENTED** [review]
whoops forgot to click "submit"

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 1]
```diff
@@ -1278,6 +1284,76 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
+        let mut finder = ExprFinder {
+            expr_span: span,
+            issue_span,
+            body_expr: None,
+            loop_expr: None,
+            loop_bind: None,
+        };
+        finder.visit_expr(hir.body(body_id).value);
+        let Some(_body_expr) = finder.body_expr else { return };
+        let Some(_loop_expr) = finder.loop_expr else { return };
```
[the remark this answers is not in the document]
Comment (chenyukang) [reply compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs]: they are left after some code refactoring...
yes, changed it to boolean would be ok.

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 2]
```diff
@@ -1278,6 +1284,76 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
+                    let hir::PatKind::Struct(path, [field, ..], _) = bind.pat.kind &&
+                    let hir::QPath::LangItem(LangItem::OptionSome, _, _) = path &&
+                    let PatField { pat: hir::Pat{ kind: hir::PatKind::Binding(_, _, ident, ..), .. }, ..} = field &&
+                    self.issue_span.source_equal(call.span) {
+                        self.loop_expr = Some(call);
+                        self.loop_bind = Some(ident.name);
+                    }
+
+                if let hir::ExprKind::MethodCall(body_call, body_receiver, _, _) = ex.kind &&
+                    body_call.ident.name == sym::next &&
```
[the remark this answers is not in the document]
Comment (chenyukang) [reply compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs]: `Iterator::next` is not a diagnostic item, but we can use typecheck result to double confirm the `trait_did` is an `Iterator`, updated the code.

**Review by compiler-errors — APPROVED** [review]
r=me if ci comes back green

**Pull request merged**